Support Android M permissions

Android M has new permission model. We need to support this.

From what I can tell this only affects the GPS location support.

More info in https://developer.android.com/preview/features/runtime-permissions.html

Currently in the Android M simulator the test app crashes as soon as you hit the GPS button.

We need to bring (or refactor) the work @ljbade did in https://github.com/mapbox/mapbox-gl-native/issues/2112 to `MyLocationTrackingModeActivity`. We include the right permissions in the manifest, but we need to check for them [at runtime](http://developer.android.com/training/permissions/requesting.html).

Right now, on an Android 6.0 device, the TestApp crashes with the following stack trace:

```
         AndroidRuntime  D  Shutting down VM
                         E  FATAL EXCEPTION: main
                         E  Process: com.mapbox.mapboxgl.testapp, PID: 7505
                         E  java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.mapboxgl.testapp/com.mapbox.mapboxsdk.testapp.MyLocationTrackingModeActivity}: java.lang.Securi
                            tyException: "passive" location provider requires ACCESS_FINE_LOCATION permission.
                         E      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
                         E      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                         E      at android.app.ActivityThread.-wrap11(ActivityThread.java)
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                         E      at android.os.Handler.dispatchMessage(Handler.java:102)
                         E      at android.os.Looper.loop(Looper.java:148)
                         E      at android.app.ActivityThread.main(ActivityThread.java:5417)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                         E  Caused by: java.lang.SecurityException: "passive" location provider requires ACCESS_FINE_LOCATION permission.
                         E      at android.os.Parcel.readException(Parcel.java:1599)
                         E      at android.os.Parcel.readException(Parcel.java:1552)
                         E      at android.location.ILocationManager$Stub$Proxy.getLastLocation(ILocationManager.java:717)
                         E      at android.location.LocationManager.getLastKnownLocation(LocationManager.java:1200)
                         E      at com.mapzen.android.lost.internal.FusionEngine.getLastLocation(FusionEngine.java:50)
                         E      at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl.getLastLocation(FusedLocationProviderApiImpl.java:32)
                         E      at com.mapbox.mapboxsdk.views.UserLocationView.toggleGps(UserLocationView.java:335)
                         E      at com.mapbox.mapboxsdk.views.UserLocationView.setEnabled(UserLocationView.java:261)
                         E      at com.mapbox.mapboxsdk.views.MapView.setMyLocationEnabled(MapView.java:3188)
                         E      at com.mapbox.mapboxsdk.testapp.MyLocationTrackingModeActivity.onCreate(MyLocationTrackingModeActivity.java:61)
                         E      at android.app.Activity.performCreate(Activity.java:6237)
                         E      at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
                         E      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
                         E      ... 9 more
        CrashlyticsCore  I  Crashlytics report upload complete: 563A37D1008B-0001-1D51-4F70B9451C7C.cls
                Process  I  Sending signal. PID: 7505 SIG: 9

```

--- a/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/MainActivity.java
+++ b/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/MainActivity.java
@@ -59,6 +59,7 @@ public class MainActivity extends AppCompatActivity {
 
     // Used for permissions requests
     private static final int PERMISSIONS_LOCATION = 0;
+    private static final int PERMISSIONS_TRACKING_MODE_ACTIVITY = 1;
 
     // Used for info window
     private static final DecimalFormat LAT_LON_FORMATTER = new DecimalFormat("#.#####");
@@ -217,6 +218,14 @@ public class MainActivity extends AppCompatActivity {
     protected void onStart() {
         super.onStart();
 
+        // We need to recheck permissions in case user revoked them via settings app
+        if ((ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
+                != PackageManager.PERMISSION_GRANTED) ||
+                (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+                        != PackageManager.PERMISSION_GRANTED)) {
+            toggleGps(false);
+        }
+
         mMapView.onStart();
     }
 
@@ -288,11 +297,17 @@ public class MainActivity extends AppCompatActivity {
     @Override
     public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
         switch (requestCode) {
-            case PERMISSIONS_LOCATION: {
-                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            case PERMISSIONS_LOCATION:
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                     enableGps();
                 }
-            }
+                break;
+
+            case PERMISSIONS_TRACKING_MODE_ACTIVITY:
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    startActivity(new Intent(getApplicationContext(), MyLocationTrackingModeActivity.class));
+                }
+                break;
         }
     }
 
@@ -347,7 +362,18 @@ public class MainActivity extends AppCompatActivity {
                                 return true;
 
                             case R.id.action_user_tracking_mode:
-                                startActivity(new Intent(getApplicationContext(), MyLocationTrackingModeActivity.class));
+                                if ((ContextCompat.checkSelfPermission(MainActivity.this, Manifest.permission.ACCESS_COARSE_LOCATION)
+                                        != PackageManager.PERMISSION_GRANTED) ||
+                                        (ContextCompat.checkSelfPermission(MainActivity.this, Manifest.permission.ACCESS_FINE_LOCATION)
+                                                != PackageManager.PERMISSION_GRANTED)) {
+                                    ActivityCompat.requestPermissions(MainActivity.this,
+                                            new String[]{Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION},
+                                            PERMISSIONS_TRACKING_MODE_ACTIVITY);
+                                } else {
+                                    startActivity(new Intent(getApplicationContext(), MyLocationTrackingModeActivity.class));
+                                }
+
+
                                 return true;
 
                             case R.id.action_polyline:

